File: tools/redis/c/filemapping.cpp
#include "filemapping.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <string>

#define LogPrintf printf

using namespace std;

const string FileMapping::dsPath_ = "/tmp/redis.sock";
const string FileMapping::globalFileIdKey_ = "global:nextFileId";

const struct timeval FileMapping::connTimeout_ = { 1, 500000 };

FileMapping::FileMapping() :
    rContext_(NULL)
{
}

FileMapping::~FileMapping()
{
    if(rContext_)
        redisFree(rContext_);
}

bool FileMapping::Init()
{
    if(!rContext_)
    {
        rContext_ = redisConnectUnixWithTimeout(dsPath_.c_str(), connTimeout_);
    }

    if(!rContext_)
    {
        LogPrintf("initliaze failure - connect to redis fail\n");
        return false;
    }

    return true;
}

/*
 * Add a *NEW* File object
 * it doesn't check the existence of lpath, just create a new one.
 *
 */
bool FileMapping::AddFile(const string &lpath, const string &rpath)
{
    fid_t fid = 0;
    fid_t pvfid = 0;
    string parentLPath;
    bool isParentVnodeFound = false;

    fid = getNextFileId();

    if(!isRootDir(lpath))
    {
        parentLPath = lpath.substr(0, lpath.find_last_of("/"));

        isParentVnodeFound = isLPathVnodeExist(parentLPath, pvfid);

        if(!isParentVnodeFound)
        {
            AddVnode(parentLPath, fid);
        } else {
            AddVnodeChild(pvfid, fid);
        }
    }

    AddFileNode(lpath, rpath, fid);

    return true;
}

/*
 * Get the RPATH of a file
 * TODO the path sperator can also be "\"
 *
 */
bool FileMapping::GetRPath(const string &lpath, SearchResult &result)
{
    bool isFound = false;
    fid_t fid = 0;
    string tmpStr("");

    isFound = isLPathNodeExist(lpath, fid);
    result.SetExactFound(isFound);
    if(isFound)
    {
        tmpStr = getRPathFromFid(fid);
        result.SetExactRPath(tmpStr);
        return true;
    }

    size_t lpos = lpath.find("/", 2);
    if(lpos == string::npos)
    {
        LogPrintf("something is wrong");
        // TODO throw
        return false;
    }

    size_t pos;
    string leftStr = lpath;
    string rightStr = "";
    pos = leftStr.find_last_of("/");

    while(pos != lpos)
    {
        rightStr = lpath.substr(pos);
        leftStr = leftStr.substr(0, pos);

        isFound = isLPathNodeExist(leftStr, fid);

        if(isFound)
        {
            tmpStr = getRPathFromFid(fid);
            result.SetParentFound(true);
            result.SetParentRPath(tmpStr);
            tmpStr += rightStr;
            result.SetExactRPath(tmpStr);
            break;
        } else
        {
            pos = leftStr.find_last_of("/");
        }
    }

    if(!isFound)
    {
        result.SetParentFound(false);
    }

    return true;
}

string FileMapping::getRPathFromFid(const fid_t fid)
{
    checkRedisContext();

    redisReply *reply;
    string rtn("");

    string tmpStr = getFileRPathKey(fid);
    reply = (redisReply*) redisCommand(rContext_, "GET %s", tmpStr.c_str());

    if(reply->type == REDIS_REPLY_ERROR ||
            reply->type == REDIS_REPLY_NIL)
    {
        // TODO throw
        return rtn;
    }

    rtn = reply->str;

    return rtn;
}

bool FileMapping::checkRedisContext()
{
    // TODO check the redis context before calling
    return true;
}

bool FileMapping::AddVnodeChild(const fid_t pvfid, const fid_t fid)
{
    checkRedisContext();

    redisReply* reply;
    string tmpStr = getFileChildrenKey(pvfid);

    reply = (redisReply*) redisCommand(rContext_, "SADD %s %lld", tmpStr.c_str(), fid);

    freeReplyObject(reply);

    return true;
}

bool FileMapping::AddFileNode(const string &lpath, const string &rpath,
                                fid_t fid)
{
    checkRedisContext();

    redisReply *reply;
    int count = 0;
    string tmpStr;
    string command;

    tmpStr = getFileLPathKey(fid);
    command = "SET " + tmpStr + " " + lpath;
    redisAppendCommand(rContext_, command.c_str());
    count++;

    tmpStr = getFileRPathKey(fid);
    command = "SET " + tmpStr + " " + rpath;
    redisAppendCommand(rContext_, command.c_str());
    count++;

    tmpStr = getLPathFidKey(lpath);
    command = "SET " + tmpStr + " " + fidToString(fid);
    redisAppendCommand(rContext_, command.c_str());
    count++;

    for(int i = 0; i < count; i++)
    {
        redisGetReply(rContext_, (void**) &reply);
        freeReplyObject(reply);
    }

    return true;
}

bool FileMapping::AddVnode(const string &lpath, fid_t childFid)
{
    checkRedisContext();

    redisReply *reply;
    fid_t fid = getNextFileId();
    int count = 0;

    string tmpStr;
    string command;

    tmpStr = getFileChildrenKey(fid);
    command = "SADD " + tmpStr + " " + fidToString(childFid);
    redisAppendCommand(rContext_, command.c_str());
    count++;

    tmpStr = getLPathVFidKey(lpath);
    command = "SET " + tmpStr + " " + fidToString(fid);
    redisAppendCommand(rContext_, command.c_str());
    count++;

    for(int i = 0; i < count; i++)
    {
        redisGetReply(rContext_, (void**) &reply);
        freeReplyObject(reply);
    }

    return true;
}

bool FileMapping::isLPathNodeExist(const string &lpath, fid_t &fid)
{
    checkRedisContext();

    string lpathFidKey = getLPathFidKey(lpath);
    redisReply *reply;

    reply = (redisReply*) redisCommand(rContext_, "GET %s", lpathFidKey.c_str());

    if(reply->type == REDIS_REPLY_ERROR)
    {
        // TODO throw
        LogPrintf("redis command failed\n");
        return false;
    } else
    {
        if(reply->type == REDIS_REPLY_NIL)
        {
            // no key found
            return false;
        }

        fid = atoi(reply->str);
        return true;
    }
}

bool FileMapping::isLPathVnodeExist(const string &lpath, fid_t &fid)
{
    checkRedisContext();

    string lpathVFidKey = getLPathVFidKey(lpath);
    redisReply *reply;

    reply = (redisReply*) redisCommand(rContext_, "GET %s", lpathVFidKey.c_str());

    if(reply->type == REDIS_REPLY_ERROR)
    {
        // TODO throw
        LogPrintf("redis command failed\n");
        return false;
    } else
    {
        if(reply->type == REDIS_REPLY_NIL)
        {
            // no key found
            return false;
        }

        fid = atoi(reply->str);
        return true;
    }
}

fid_t FileMapping::getNextFileId()
{
    checkRedisContext();

    redisReply *reply;
    fid_t rtn = -1;

    reply = (redisReply*) redisCommand(rContext_, "INCR %s", globalFileIdKey_.c_str());
    if(reply->type == REDIS_REPLY_ERROR)
    {
        // TODO throw exception here
        return rtn;
    }
    else
    {
        rtn = reply->integer;
        freeReplyObject(reply);

        return rtn;
    }
}

bool FileMapping::isRootDir(const string &lpath)
{
    string str = lpath.substr(2);
    size_t pos;

    pos = str.find("/");
    str = str.substr(pos+1);

    if(str.find("/") == string::npos)
        return true;
    else
        return false;
}

string FileMapping::getLPathFidKey(const string &lpath)
{
    return "lpath:" + lpath + ":fid";
}

string FileMapping::getLPathVFidKey(const string &lpath)
{
    return "lpath:" + lpath + ":vfid";
}

string FileMapping::getFileTypeKey(const fid_t fid)
{
    return "file:" + fidToString(fid) + ":type";
}

string FileMapping::getFileChildrenKey(const fid_t fid)
{
    return "file:" + fidToString(fid) + ":children";
}

string FileMapping::getFileLPathKey(const fid_t fid)
{
    return "file:" + fidToString(fid) + ":lpath";
}

string FileMapping::getFileRPathKey(const fid_t fid)
{
    return "file:" + fidToString(fid) + ":rpath";
}

string FileMapping::fidToString(const fid_t fid)
{
    char buf[128];

    memset(buf, 0, sizeof(buf));
    sprintf(buf, "%lld", fid);

    string tmpStr(buf);

    return tmpStr;
}

File: tools/redis/nodejs/filemapping.js
var redis = require("redis");
var async = require("async");

var client = redis.createClient();

var global_next_fileid = "global:nextFileId";

module.exports = filemapping;

client.on("error", function(err) {
    console.log("redis client error: " + err);
});

client.on("end", function() {
    console.log("redis server close conn, reconnect ..");
    client = redis.createClient();
});

function filemapping() {
}

filemapping.prototype.addfile = function(lpath, rpath) {
    var fid;
    var pfid, plpath;
    var is_parent_exist = false;
    var is_vnode_found = false;

    var is_root = is_root_dir(lpath);
    var plpath = lpath.substring(0, lpath.lastIndexOf("/"));

    var multi = client.multi();

    async.series([
        function(callback) {
            client.incr(global_next_fileid, function(err, result) {
                fid = result;
                callback(null, 1);
            });
        },
        function(callback) {
            if(!is_root) {
                client.exists(get_lpath_set_key(plpath), function(err, result) {
                    if(err) {
                        console.log("client exists error", err);
                        return;
                    }

                    if(result == 1)
                        is_parent_exist = true;

                    callback(null, 1);
                });
            }
            callback(null, 1);
        },
        function(callback) {
            if(!is_root && (!is_parent_exist || !is_vnode_found)) {
                client.incr(global_next_fileid, function(err, result) {
                    if(err) { console.log("Error", err); }
                    pfid = result;
                    callback(null, 1);
                });
            }

            callback(null, 1);
        },
        function(callback) {
            if(!is_root && is_parent_exist) {
                client.smembers(get_lpath_set_key(plpath), function(err, pfids) {
                    pfids.forEach(function(id, i) {
                        if(is_vnode(id)) {
                            multi.sadd(get_file_children_key(id), fid);
                            is_vnode_found = true;
                        }
                    });
                    callback(null, 1);
                });
            }
            callback(null, 1);
        },
        function(callback) {
            if(!is_root && (!is_parent_exist || !is_vnode_found)) {
                multi.set(get_file_type_key(pfid), 0);
                multi.sadd(get_file_children_key(pfid), fid);
                multi.sadd(get_lpath_set_key(plpath), pfid);
            }

            multi.set(get_file_lpath_key(fid), lpath);
            multi.set(get_file_rpath_key(fid), rpath);
            multi.sadd(get_lpath_set_key(lpath), fid);

            multi.exec(function(err, replies) {
                console.log(replies);
            });
        }
    ], function(err, result) {
        if(err) console.log("Final Error:", err);
        console.log("add file succeed");
        client.quit();
    });
};

function is_vnode(fid)
{
    var rtn = false;
    var done = false;
    client.exists(get_file_type_key(fid), function(err, result) {
        if(err) {
            console.log("client exist error", err);
            return;
        }

        rtn = (result == 1) ? true : false;
        done = true;
    });

    while(!done) {}

    return rtn;
}

function get_new_fileid()
{
    var fid;
    var done = false;

    client.incr(global_next_fileid, function(err, result) {
        fid = result;
        done = true;
    });

    return fid;
}

// assume file lpath format is:
//   == //host/home/iwss ==
function is_root_dir(lpath)
{
    var str = lpath.substring(2);
    str = str.substring(str.indexOf("/") + 1);

    if(str.indexOf("/") == -1)
        return true;
    else
        return false;
}

function get_file_lpath_key(fid)
{
    var rtn = "file:" + fid + ":lpath";
    return rtn;
}

function get_file_rpath_key(fid)
{
    return "file:" + fid + ":rpath";
}

function get_lpath_set_key(lpath)
{
    return "lpath:" + lpath + ":fid";
}

function get_file_type_key(fid)
{
    return "file:" + fid + ":type";
}

function get_file_children_key(fid)
{
    return "file:" + fid + ":children";
}
